glib/tests/clone.rs

```diff
@@ -457,23 +457,33 @@ fn test_clone_macro_default_return() {
 
     let _closure = clone!(@weak v => @default-return Foo(0), move || Foo(1));
 
-    #[allow(dead_code)]
+    #[derive(PartialEq, Debug)]
     struct Bar {
         x: i32,
     }
 
-    let _closure = clone!(@weak v => @default-return Bar { x: 0 }, move || Bar { x: 1 });
+    let closure = clone!(@weak v => @default-return Bar { x: 0 }, move || Bar { x: 1 });
+    let wrapper = clone!(@weak v => move |call: &(dyn Fn() -> Bar)| call());
+    assert_eq!(wrapper(&closure), Bar { x: 0 });
```

Comment:
This test is failing because it seems like it can upgrade `v` twice, which is very weird...

Reply:
Can you provide the `cargo expand` output for the failing tests?

Reply:
Enjoy! So this is `closure`:

```rust
{
    let v = glib::clone::Downgrade::downgrade(&v);
    move || {
        let v = match glib::clone::Upgrade::upgrade(&v) {
            Some(val) => val,
            None => {
                glib::g_debug!(glib::CLONE_MACRO_LOG_DOMAIN, "Failed to upgrade v",);
                let ___return_value = || Bar { x: 0 };
                return ___return_value();
            }
        };
        {
            Bar { x: 1 }
        }
    }
}
```

And this is `wrapper`:

```
{
    let v = glib::clone::Downgrade::downgrade(&v);
    move |call: &(dyn Fn() -> Bar)| {
        let v = match glib::clone::Upgrade::upgrade(&v) {
            Some(val) => val,
            None => {
                panic!("failed to upgrade `v` (if you don't want to panic, use @default-return)",)
            }
        };
        {
            call()
        }
    }
}
```

Reply:
What's the problem here? I would expect it to return `Bar { x: 1 }` here and not `Bar { x: 0 }`. Why do you expect the other?

Reply:
The problem was me. :)